recipes-extended/xen/files/libxl-blktap3-support.patch

```diff
@@ -0,0 +1,703 @@
+https://github.com/qiaohaiming/blktap3/blob/master/libxl.diff
+
+Contributions by:
+Mahantesh Salimath <salimathm@ainfosec.com>
+
+################################################################################
+UPSTREAM PLAN
+################################################################################
+Could be upstreamed
+
+################################################################################
+PATCHES
+################################################################################
+Index: xen-4.9.1/tools/libxl/libxl_blktap3.c
+===================================================================
+--- /dev/null
++++ xen-4.9.1/tools/libxl/libxl_blktap3.c
+@@ -0,0 +1,139 @@
++/*
++ * Copyright (C) 2012      Advanced Micro Devices
++ * Author Christoph Egger <Christoph.Egger@amd.com>
++ *
++ * This program is free software; you can redistribute it and/or modify
++ * it under the terms of the GNU Lesser General Public License as published
++ * by the Free Software Foundation; version 2.1 only.
++ *
++ * This program is distributed in the hope that it will be useful,
++ * but WITHOUT ANY WARRANTY; without even the implied warranty of
++ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
++ * GNU Lesser General Public License for more details.
++ */
++
++#include "libxl_osdeps.h" /* must come before any other headers */
++#include "libxl_internal.h"
++
++#include <tap-ctl.h>
++
++#include "list.h"   /* include for list_head structure */
++
++static int blktap_find(const char *type, const char *path, struct tap_list *tap)
++{
++    struct list_head list; /* Note: structure name updated */
++    struct tap_list *entry, *next_t;
++    int ret = -ENOENT, err;
++
++    /* TAILQ_INIT(&list);--> old function */
++    INIT_LIST_HEAD(&list);
++
++    if ((err = tap_ctl_list(&list)))
++        return err;
++
++    /* TAILQ_EMPTY(&list)--> old function */
++    if (list_empty(&list))
++        return ret;
++
++    tap_list_for_each_entry_safe(entry, next_t, &list) {
++
++        if (type && (!entry->type || strcmp(entry->type, type)))
++            continue;
++
++        if (path && (!entry->path || strcmp(entry->path, path)))
++            continue;
++
++        *tap = *entry;
++	tap->type = tap->path = NULL;
++        ret = 0;
++        break;
++    }
++
++    tap_ctl_list_free(&list);
++
++    return ret;
++}
++
++/**
++ * blktap3 doesn't require blkback, so it's always available.
++ */
++int libxl__blktap_enabled(libxl__gc *gc)
++{
++    return 1;
++}
++
++char *libxl__blktap_devpath(libxl__gc *gc, const char *disk,
++		libxl_disk_format format)
++{
++    const char *type = NULL;
++    char *params, *devname = NULL;
++    struct tap_list tap;
++    int err = 0;
++    int flags = 0;
++
++    type = libxl__device_disk_string_of_format(format);
++    if (!(err = blktap_find(type, disk, &tap))) {
++        LOG(DEBUG, "found tapdisk\n");
++	devname = libxl__sprintf(gc, "/dev/xen/blktap-2/tapdev%d", tap.minor);
```

Comment:
Are we keeping this devpath as blktap-2 for a reason?

Reply:
Just that the blktap3 repository from xenserver (https://github.com/xapi-project/blktap) has not changed it, new tapdev is written to /dev/xen/blktap-2/ devpath. Since the base for my changes is xenserver repository I didn't change it. Let me know if this is okay or needs to be changed.